src/utils/simplifier.js

```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
   switch (value.kind) {
     case "!":
       return negate(realm, value.args[0]);
+    case "||":
+    case "&&":
+      let x = simplify(realm, value.args[0]);
+      let y = simplify(realm, value.args[1]);
+      if (x instanceof AbstractValue && x.equals(y)) return x;
+      return AbstractValue.createFromLogicalOp(realm, (value.kind: any), x, y, value.expressionLocation);
+    case "==":
+    case "!=":
+    case "===":
+    case "!==":
+      return simplifyEquality(realm, value);
     default:
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (y instanceof AbstractValue) [x, y] = [y, x];
```

Comment:
You could reduce unneeded swapping by testing for `!(x instanceof AbstractValue)`, or `x instanceof ConcreteValue`.

Reply:
Fair enough. Will change.